crates/store/src/state.rs

```diff
@@ -653,25 +737,101 @@ impl State {
 
     /// Lists all known nullifiers with their inclusion blocks, intended for testing.
     pub async fn list_nullifiers(&self) -> Result<Vec<(Nullifier, u32)>, DatabaseError> {
-        self.db.select_nullifiers().await
+        self.db.select_all_nullifiers().await
     }
 
     /// Lists all known accounts, with their ids, latest state hash, and block at which the account
     /// was last modified, intended for testing.
     pub async fn list_accounts(&self) -> Result<Vec<AccountInfo>, DatabaseError> {
-        self.db.select_accounts().await
+        self.db.select_all_accounts().await
     }
 
     /// Lists all known notes, intended for testing.
     pub async fn list_notes(&self) -> Result<Vec<NoteRecord>, DatabaseError> {
-        self.db.select_notes().await
+        self.db.select_all_notes().await
     }
 
-    /// Returns details for public (public) account.
+    /// Returns details for public (on-chain) account.
     pub async fn get_account_details(&self, id: AccountId) -> Result<AccountInfo, DatabaseError> {
         self.db.select_account(id).await
     }
 
+    /// Returns account states with details for public accounts.
+    pub async fn get_account_states(
+        &self,
+        requests: &HashMap<AccountId, AccountStateRequest>,
+    ) -> Result<(BlockNumber, Vec<AccountStateResponse>), DatabaseError> {
+        'reload: loop {
+            let (snapshot_block_num, proofs) = {
+                let inner_state = self.inner.read().await;
+                let mut proofs = HashMap::new();
+                for account_id in requests.keys() {
+                    let key = LeafIndex::new_max_depth(*account_id);
+                    let proof = inner_state.account_tree.open(&key);
+
+                    proofs.insert(*account_id, proof);
+                }
+
+                (Self::compute_latest_block_number(&inner_state), proofs)
+            };
+
+            let infos = self.db.select_accounts_by_ids(requests.keys().copied().collect()).await?;
+
+            // TODO: Should we return an error if one or more accounts are not found?
```

Comment:
Yes, we can tell just from the account ID if it is public or not ([see here](https://github.com/0xPolygonMiden/miden-base/blob/next/objects/src/accounts/account_id.rs#L249)). There are two things we can do about private accounts:

1. Return only the inclusion path.
2. Return an error.

I think for now, we should go with the latter option and we can change this in future if needed. So, basically, this endpoint becomes exclusively for public accounts (and maybe we should name it accordingly? - e.g., `GetPublicAccountStates`).

Also, since we can make the check based on IDs - we should do it this in the RPC component.

Reply:
Another option could be to simplify this endpoint and make it work primarily to get account inclusion paths + basic information for accounts (when available). And this would include both public and private accounts. So, for example, the response could be (other naming suggestions welcome):
```proto
message AccountStateResponse {
    account.AccountId account_id = 1;
    digest.Digest account_hash = 2;
    merkle.MerklePath account_proof = 3;
    optional AccountStateHeader = 4;
}

message AccountStateHeader {
    account.AccountHeader header = 1;
    bytes storage_header = 2;
}
```
This would cover 90% of what we need (if not more), and for additional data (e.g., storage maps and assets) we could have a different endpoint (maybe we could just adapt our current `GetAccountDetails` endpoint for that). But this would be for another PR.

Reply:
Personally I like the simplify suggestion. Its nice having well defined endpoints that just work without exceptions.

I think the naming is fine.